components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth/action/execution/PreIssueAccessTokenRequestBuilder.java

```diff
@@ -112,12 +121,69 @@ private Event getEvent(OAuthTokenReqMessageContext tokenMessageContext, Map<Stri
             eventBuilder.userStore(new UserStore(authorizedUser.getUserStoreDomain()));
         }
 
-        eventBuilder.accessToken(getAccessToken(tokenMessageContext, claimsToAdd));
+        // Retrieving app information here to reduce the perf overhead.
+        OAuthAppDO oAuthAppDO = getAppInformation(tokenMessageContext);
+
+        eventBuilder.accessToken(getAccessToken(tokenMessageContext, oAuthAppDO, claimsToAdd));
+
+        if (isRefreshTokenAllowed(oAuthAppDO)) {
+            eventBuilder.refreshToken(getRefreshToken(oAuthAppDO, tokenMessageContext));
+        }
         eventBuilder.request(getRequest(tokenReqDTO));
 
         return eventBuilder.build();
     }
 
+    private boolean isRefreshTokenAllowed(OAuthAppDO oAuthAppDO) {
+
+        if (OAuthServerConfiguration.getInstance().getSupportedGrantTypes().containsKey(
+                GrantType.REFRESH_TOKEN.toString()) && oAuthAppDO != null) {
+
+            String grantTypes = oAuthAppDO.getGrantTypes();
+            if (StringUtils.isNotEmpty(grantTypes)) {
+                List<String> supportedGrantTypes = Arrays.asList(grantTypes.split(" "));
+                return supportedGrantTypes.contains(OAuthConstants.GrantTypes.REFRESH_TOKEN);
+            }
+        }
+        return false;
+    }
+
+    private RefreshToken getRefreshToken(OAuthAppDO oAuthAppDO, OAuthTokenReqMessageContext tokenMessageContext) {
+
+        long refreshTokenValidityPeriod = resolveRefreshTokenValidityPeriod(oAuthAppDO, tokenMessageContext);
+        long refreshTokenIssuedAt = resolveRefreshTokenIssuedAt(tokenMessageContext);
+
+        RefreshToken.Builder refreshTokenBuilder = new RefreshToken.Builder()
+                .addClaim(RefreshToken.ClaimNames.EXPIRES_IN.getName(), refreshTokenValidityPeriod);
+
+        // Adding this claim to improve the user experience only when existing refresh token is reused.
```

Comment:
This comment is also redundant. we don't need to explain functionality in code

Reply:
Fixed 91b6ce9040c88be6f1ba11fbd755a76e26411be6